src/obj_dir.mli

```diff
@@ -1,73 +1,85 @@
+    directory contains several libraries and/or executables.
+
+    This also allows us to provide a different API for a given
+    library.  In particular, depending on the context we might choose
+    to expose or not the private modules.
+
+    In the rest of this API, "local" and "external" have their usual
+    Dune meaning: "local" is for libraries or executables that are
+    local to the current worksapce and "extenal" for libraries that are
+    part of the installed world.
+
+    For local libraries, the path are reported as [Path.Build.t]
+    values given that they are all inside the build directory.  For
+    external libraries the path are reported as [Path.t]
+    values. However, it is possible to get a view of the object
+    directory for a local library where the path are reported as
+    [Path.t] values with [of_local].  This is convenient in places
+    where we need to treat object directories of both local and
+    external library in the same way.
+*)
 
-  val all_obj_dirs : t -> mode:Mode.t -> Path.Build.t list
-
-  (** The public compiled cmi file directory *)
-  val public_cmi_dir: t -> Path.Build.t
-
-  val make_lib
-    :  dir:Path.Build.t
-    -> has_private_modules:bool
-    -> Lib_name.Local.t
-    -> t
-end
+open! Stdune
 
-type t
+type 'path t
 
-val of_local : Local.t -> t
+val of_local : Path.Build.t t -> Path.t t
 
 (** The source_root directory *)
-val dir : t -> Path.t
+val dir : 'path t -> 'path
 
 (** The directory for ocamldep files *)
-val obj_dir : t -> Path.t
+val obj_dir : 'path t -> 'path
 
 (** The private compiled native file directory *)
-val native_dir : t -> Path.t
+val native_dir : 'path t -> 'path
 
 (** The private compiled byte file directories, and all cmi *)
-val byte_dir : t -> Path.t
+val byte_dir : 'path t -> 'path
 
-val all_cmis: t -> Path.t list
+val all_cmis: 'path t -> 'path list
 
 (** The public compiled cmi file directory *)
-val public_cmi_dir: t -> Path.t
-
-val pp: t Fmt.t
-val to_sexp: t -> Sexp.t
+val public_cmi_dir: 'path t -> 'path
 
-val all_obj_dirs : t -> mode:Mode.t -> Path.t list
+val all_obj_dirs : 'path t -> mode:Mode.t -> 'path list
 
+(** Create the object directory for a library *)
 val make_lib
   :  dir:Path.Build.t
   -> has_private_modules:bool
   -> Lib_name.Local.t
-  -> t
+  -> Path.Build.t t
+
+(** Create the object directory for a set of executables. [name] is
+    name of one of the executable in set. It is included in the dot
+    subdirectory name. *)
+val make_exe: dir:Path.Build.t -> name:string -> Path.Build.t t
+
+(** Create the object directory for an external library that has no
+   private directory for private modules *)
+val make_external_no_private : dir:Path.t -> Path.t t
+
+val encode : Path.t t -> Dune_lang.t list
+val decode : dir:Path.t -> Path.t t Dune_lang.Decoder.t
 
-val make_external_no_private : dir:Path.t -> t
+val convert_to_external : Path.t t -> dir:Path.t -> Path.t t
```

Comment:
Why doesn't this take a `Path.Build.t`?

Reply:
The place where this is used is when convert `Lib.t` values to to `Dune_package.Lib.t` values. This is where the layout of the object directory changes. Our code technically accepts any `Lib.t` value here, but really, only `Lib.Local.t` values can (or should) be converted to dune package libraries.

So to make this have the intuitive type signature as you've observed we need to  have `to_dune_lib` take a `Lib.Local.t`. I believe this is outside of the scope of this PR.

By the way, we can also paramterize `Lib.t` in a similar way as well. So that `Lib.Local.t = Path.Build.t Lib.Local.t` and we'll avoid duplicating the API in 2 modules: Lib and Lib.Local

Reply:
I see. However, when looking only at this API, it's hard to understand why this choice was made.

What about changing this function to take a `Path.Build.t t` and adding another function `to_local : Path.t t -> Path.Build.t t option`?

Reply:
> By the way, we can also paramterize Lib.t in a similar way as well. So that Lib.Local.t = Path.Build.t Lib.Local.t and we'll avoid duplicating the API in 2 modules: Lib and Lib.Local

Ok. That seems worth a try

Reply:
I followed your suggestion for now. I will experiment with the refactoring I just mentioned in a subsequent PR.